Seccompiler minimal implementation

alindima opened the pull request:
## Reason for This PR

PR based on the proposal opened on this RFC: #2209 
Implements the seccompiler functionality and the accommodating changes in Firecracker.

## Description of Changes

- Modify Firecracker to use thread-specific seccomp filters

- Replace `--seccomp-level` Firecracker param with `--no-seccomp` and `--seccomp-filter`
- Implemented the seccompiler binary, which compiles JSON filters into serialized binary BPF.
- Implemented the custom filter functionality in Firecracker.
- Added python integration tests for the new seccomp parameters.
- Rewritten python integration_tests in `test_seccomp.py` to use and test seccompiler.
- Removed Firecracker's uses of `seccomp/lib` in rust tests.
- Added `devtool generate_syscall_tables` command

_**For a greater perspective, these things are yet to be implemented in further PRs**_:
- _writing the Firecracker JSON filters_
- _build-time mechanism for embedding the compiled filters in Firecracker_
- _firecracker documentation on seccomp_
- _seccompiler documentation_

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

iulianbarbu commented on src/seccomp/src/bin.rs:
```diff
@@ -0,0 +1,567 @@
+        )
+}
+
+fn get_argument_values(arguments: &ArgumentsBag) -> Result<Arguments> {
```
This can be as well a `TryInto` implementation or a `new` constructor for a struct which encapsulates all the logic. WDYT?

alindima replied:
hmm, I think that the added complexity would not be justified. I feel that the procedural approach is more clear here

iulianbarbu commented on src/seccomp/src/bin.rs:
```diff
@@ -0,0 +1,567 @@
+            )
+            .unwrap();
+        assert!(get_argument_values(arguments).is_err());
+        // invalid --target_arch
```
You have a duplicate case of invalid `--target-arch` here.

alindima replied:
will remove it

iulianbarbu commented on src/seccomp/src/bin.rs:
```diff
@@ -0,0 +1,567 @@
+            }
+        );
+
+        let arguments = &mut arg_parser.arguments().clone();
```
I would also add cases when you do not provide values to `--input_file`, `--target-arch` and `--output_file` as well. WDYT?

alindima replied:
> I would also add cases when you do not provide values to --input_file, --target-arch and --output_file as well. WDYT?

I didn't provide test cases for these since they are either `required` or have default argument values. However, I will add all the above test cases, and will try to think of some more. Would be nice to have the extra assurance

alindima replied:
added test cases for: missing `--input-file`, missing `--target-arch` and inexistent input file

iulianbarbu commented on src/seccomp/src/bin.rs:
```diff
@@ -0,0 +1,567 @@
+    use utils::tempfile::TempFile;
+
+    // test helper for generating correct JSON input data
+    fn get_correct_json_input() -> String {
```
```suggestion
    fn default_json_input() -> String {
```

alindima replied:
hmm, this input is in no way default. It's just an example of a correct filter file

iulianbarbu reviewed (commented):
First part of the review. Will continue later for the rest.

alindima commented on .gitignore:
```diff
@@ -5,3 +5,4 @@ target
 .idea
 __pycache__
 *.pyc
+*.core
```
sometimes, when you force-quit a devtool test run, `core.*` files seem to be generated

alindima commented on src/firecracker/src/api_server_adapter.rs:
```diff
@@ -121,7 +121,7 @@ pub fn run_with_api(
         .expect("Failed to clone API event FD");
 
-    let api_seccomp_filter = seccomp_filter.clone();
+    let api_seccomp_filter = seccomp_filters.remove("Api").unwrap();
```
> Does this mean the seccomp_filters will always contain the "Api" thread seccomp filtering information, either specified by users or by Firecracker, through a default seccomp filter?
Yes. This check is done on the `get_custom_filter` method in the `default_syscalls` module.
I agree that `expect` would be better for this case.

alindima replied:
In regards to the names of the thread categories, I prefer `Api`, `Vmm` and `Vcpu`. It would be redundant to embed the `fc_*` prefix in the naming scheme. Also, `firecracker` instead of `Vmm` may be confusing, since all threads belong to the Firecracker process.
Naming things is hard 😄

alindima commented on src/firecracker/src/main.rs:
```diff
@@ -246,8 +257,12 @@ fn main() {
+            .collect();
         run_without_api(
-            seccomp_filter,
+            &mut seccomp_filters,
```
yes, because further down the line, we will call `remove` on the map, in order to get ownership of the mapped filter

alindima commented on src/seccomp/src/bin.rs:
```diff
@@ -0,0 +1,567 @@
+
+use bincode::Error as BincodeError;
+use compiler::{Compiler, Error as FilterFormatError, Filter};
+pub use seccomp::BpfThreadMap;
```
seems to be an error

alindima replied:
an error on my part*

alindima replied:
solved

alindima commented on src/seccomp/src/bin.rs:
```diff
@@ -0,0 +1,567 @@
+
+fn get_argument_values(arguments: &ArgumentsBag) -> Result<Arguments> {
+    let target_arch = arguments.value_as_string("target_arch").and_then(|val| {
+        if !SUPPORTED_ARCHES.contains(&val.as_ref()) {
+            None
+        } else {
+            Some(val)
+        }
+    });
+
+    if target_arch.is_none() {
+        return Err(Error::InvalidArch);
+    }
```
i like this, will do

alindima replied:
done. we now have a TargetArch enum that also does lowercasing for the TryInto String implementation

sandreim commented on src/seccomp/src/compiler.rs:
```diff
@@ -0,0 +1,561 @@
+#[serde(deny_unknown_fields)]
+pub(crate) struct SyscallObject {
+    /// Name of the syscall
+    syscall: Option<String>,
```
Can we drop this field? Just remove the Option from syscalls and enforce the user to always specify at least one.

alindima replied:
I actually proposed this in the initial design of the file format. However, during the design reviews, the general opinion was that we should have two distinct types of objects: 

- one that refers only to a single syscall and can have argument checks.
- anther that refers to a collection of syscalls and cannot have arguments conditions.

This arguably makes for a easier to use API, and I think this is one of the top priorities. WDYT?

alindima replied:
done

sandreim commented on src/seccomp/src/compiler.rs:
```diff
@@ -0,0 +1,561 @@
+    /// Vector of syscall names
+    syscalls: Option<Vec<String>>,
+    /// Optional on-match action
+    action: Option<SeccompAction>,
```
Is action optional ? What is the default action ?

alindima replied:
yes. the default on-match action is the `filter_action` on the parent `Filter` object

sandreim commented on src/seccomp/src/compiler.rs:
```diff
@@ -0,0 +1,561 @@
+
+    /// True if object refers to a single syscall
+    fn is_singular(&self) -> bool {
+        self.syscall.is_some() && self.syscalls.is_none()
```
This get's simpler if we remove syscall, see above.

sandreim commented on src/seccomp/src/compiler.rs:
```diff
@@ -0,0 +1,561 @@
+    /// Default action if a rule matches. e.g. `Allow` for an AllowList
+    filter_action: SeccompAction,
+    /// The collection of rules
+    filter: Vec<SyscallObject>,
```
Why not call it `rules` then?

alindima replied:
makes sense

alindima replied:
on a second thought, it doesn't make sense. I will update the comment to smth like `The collection of SyscallObjects`.

At this stage they have not yet been transformed in rules

alindima replied:
done

sandreim commented on src/seccomp/src/compiler.rs:
```diff
@@ -0,0 +1,561 @@
+            let filter_action = &filter.filter_action;
+
+            if syscall_object.is_plural() {
+                let action = syscall_object
```
This code also becomes much simpler if we  just keep syscalls.

sandreim commented on src/seccomp/src/syscall_table.rs:
```diff
@@ -0,0 +1,673 @@
+    // possible syscalls is architecture-dependent. We can't just map to their
+    // corresponding libc constant.
+    fn populate_map(&mut self) {
+        if self.arch == "aarch64" {
```
Can we  move these into separate files and just include them here?. It would make it easier to manage.

alindima replied:
makes sense

alindima replied:
done

sandreim commented on src/seccomp/src/lib.rs:
```diff
@@ -230,9 +229,11 @@
+use std::io::Read;
 
 /// Maximum number of instructions that a BPF program can have.
 const BPF_MAX_LEN: usize = 4096;
```
Are we sure we do not need to update this ?

alindima replied:
This limit is enforced by the linux kernel

sandreim reviewed (commented):
I've done a fast first pass on this PR. I will do another one in next iteration of PR.

aghecenco commented on src/seccomp/src/syscall_table.rs:
```diff
@@ -0,0 +1,673 @@
+    // corresponding libc constant.
+    fn populate_map(&mut self) {
+        if self.arch == "aarch64" {
+            self.map.insert("SYS_io_setup".to_string(), 0);
```
After giving this some more thought, I don't think this is a safe thing to do.
1. Reviewers (me :fearful:) should verify that each of these hardcoded numbers does indeed correspond to the denominated syscall. That's very error prone. Further additions to these lists are prone to the same reviewer errors: a reviewer might miss an erroneous syscall and inadvertently create a security flaw.
1. Even before the review, new additions to this list could add mistakes that will cause faulty code to run as part of the CI build. This should be avoided.

I suggest generating this list either using `build.rs` (preferred, as it protects against mistakes in future PRs) or with an external tool, then copy-pasting it here, from the header files present on the system (all of these are defined somewhere in `/usr/include`, and it's a matter of parsing C headers).

alindima replied:
putting these syscalltable modifications on hold for now. I need to think a little more about it

acatangiu replied:
Generating these seems to bring in extra complexity and security concerns or failure points. AFAIK syscall numbers are part of the kernel ABI and don't change so why not just hardcode them in a source file?

Initial commit of the hardcoded list needs to be carefully reviewed, but then changes there (which should be rare) are few-liners easy to review.

If we're really paranoid we can add extra validation like unittests that check our hardcoded values against kernel headers or smth.

alindima replied:
We synced offline and decided to auto-generate the syscall tables, on-demand. I will work on writing a script for doing this, from the linux kernel source code.

alindima replied:
done, PTAL

aghecenco reviewed (changes requested):
First pass of the review. Blocking for now on the matter of hardcoded syscall numbers.

aghecenco reviewed (commented):
Review part 2

alindima commented on src/seccomp/src/lib.rs:
```diff
@@ -515,16 +515,7 @@ impl SeccompCondition {
-                (arg_offset + SECCOMP_DATA_ARG_SIZE / 2, arg_offset)
-            }
-        };
+        let (msb_offset, lsb_offset) = { (arg_offset + SECCOMP_DATA_ARG_SIZE / 2, arg_offset) };
```
this was already introduced in a commit a little later.

alindima commented on src/seccomp/src/syscall_table.rs:
```diff
@@ -0,0 +1,673 @@
+    }
+
+    // We need to create an architecture-specific map since the set of
+    // possible syscalls is architecture-dependent. We can't just map to their
```
I agree that my comment is a bit misleading. I meant that we can't use the libc constants because at compile time, we can only get access to one target. This is because libc uses conditional compilation, while seccompiler needs access to the constants for all libc targets

alindima replied:
Also, rust libc seems to have the complete syscall table for x86_6-musl and aarch64-musl targets. Their list has been updated and I can't find any syscall that it lacks.

aghecenco replied:
That's great, and maybe we can use it :smile: in any case, I think your explanation about the compile-time restriction would make the comment less misleading.

alindima replied:
removed this comment. doesn't make that much sense now that we autogenerate the syscall tables

alindima commented on src/seccomp/src/compiler.rs:
```diff
@@ -0,0 +1,561 @@
+    #[serde(rename = "args")]
+    conditions: Option<Vec<SeccompCondition>>,
+    /// Unused field, represents a comment property in the JSON format
+    #[allow(dead_code)]
```
Since I added the `#[serde(deny_unknown_fields)]` directive on the `SyscallObject` struct, `serde_json` will complain about additional unknown fields. However, we want to allow the optional `comment` property, to specify comments.

Now that I think more about it, it doesn't make sense to store these comments in memory, and we should discard their content after deserialization.

I will introduce a new type, `Comment`, that stores no value and I will implement a custom Deserialize method for it, that simply discards the data.

alindima replied:
fixed

alindima commented on src/seccomp/src/bin.rs:
```diff
@@ -0,0 +1,567 @@
+#[derive(Debug)]
+pub enum Error {
+    Bincode(BincodeError),
+    FileOpen(PathBuf, io::Error),
```
I feel like the `FileOpen` message is more descriptive. This error can only happen on opening files, and this information can be helpful to the user.
If we rename the error to `IoError`, it would also make sense to modify the message to a general IO error, which would be less descriptive in this case.
WDYT?

alindima commented on src/seccomp/src/bin.rs:
```diff
@@ -1,8 +1,568 @@
+fn build_arg_parser() -> ArgParser<'static> {
+    ArgParser::new()
+        .arg(
+            Argument::new("input_file")
```
I have stated in the PR message that this work still needs to be done. You think it is OK to include this doc in another PR? This PR is already pretty beefy. @aghecenco ?

aghecenco replied:
Yeah, it's OK to follow up with more documentation once this is merged.

alindima commented on src/seccomp/src/bin.rs:
```diff
@@ -1,8 +1,568 @@
+        let in_file = TempFile::new().unwrap();
+        let out_file = TempFile::new().unwrap();
+
+        let mut json_input = get_correct_json_input();
```
is it fair to include more extensive JSON negative tests in a subsequent PR?

alindima replied:
Added more extensive JSON negative tests in this PR. PTAL

alindima commented on src/seccomp/src/bin.rs:
```diff
@@ -1,8 +1,568 @@
+    let bpf_data: BpfThreadMap = compiler.compile_blob(filters).map_err(Error::FileFormat)?;
+
+    // serialize the BPF programs & output them to a file
+    let output_file = File::create(&args.output_file)
```
it grants write access. No need to use OpenOptions

alindima commented on src/seccomp/src/bin.rs:
```diff
@@ -1,8 +1,568 @@
+fn compile(args: &Arguments) -> Result<()> {
+    let input_file = File::open(&args.input_file)
+        .map_err(|err| Error::FileOpen(PathBuf::from(&args.input_file), err))?;
+    let mut input_reader = BufReader::new(input_file);
```
Actually, I don't know for sure how many reads the program makes to the input file. This is in `serde_json`'s hands.
I am assuming that it does more than one `read` calls, to not bring into memory the entire file at once (the file could be huuge).
It would have been clear that it does only one `read` if seccompiler used `deserialize_str`. However, this is precisely why I chose to use `deserialize_from` (which works with Read objects and likely does more read actions)

alindima replied:
Update: 
https://docs.serde.rs/serde_json/fn.from_reader.html
`serde_json` actually recommends using BufReader for File objects.

This should also take care of OOM. WDYT?

aghecenco replied:
`from_reader` actually passes the burden of reading stuff to the `Read`er, it doesn't handle the actual reading ([here](https://github.com/serde-rs/json/blob/3d8b046d0247da886a60de0d1b1aa8ba1524cfe5/src/de.rs#L69) to [here](https://github.com/serde-rs/json/blob/3d8b046d0247da886a60de0d1b1aa8ba1524cfe5/src/read.rs#L190)); which is why I was raising the `BufReader` question, as it's indeed preferable (and recommended by the documentation) for *small* reads, but that doesn't apply to this use case, where you need to read the whole thing into memory for `serde_json` to convert it.

alindima replied:
Running seccompiler with BufReader under strace, on a file with ~600 lines of JSON, I see three calls to `read`.
In contrast, running without BufReader, seccompiler makes a call to `read` for **every single byte**.

I have also done a small benchmark test based on the `time` command, and for a ~600 lines JSON file, the BufReader approach is about ~0.01s faster. This is definitely not much, and given the fact that `seccompiler` is not really performance constrained, it really does not matter that much.

However, given the huuuge amount of syscalls, I think it would be best to keep BufReader

alindima commented on src/seccomp/src/bin.rs:
```diff
@@ -1,8 +1,568 @@
+        return Err(Error::InvalidArch);
+    }
+
+    let input_file = arguments.value_as_string("input_file");
```
the validations performed for files are different, based on whether they are input or output files.
Taking your approach would mean having a function for validating every single argument, which in my opinion is a little too verbose, WDYT?

aghecenco replied:
I don't mind verbosity but I don't mind this too much, it's more of a style nit.

alindima commented on src/seccomp/src/bin.rs:
```diff
@@ -0,0 +1,567 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
will do

alindima replied:
done

alindima commented on src/firecracker/src/main.rs:
```diff
@@ -246,8 +257,12 @@ fn main() {
     } else {
+        let mut seccomp_filters: BpfThreadMap = seccomp_filters
+            .into_iter()
+            .filter(|(k, _)| k == "Vmm" || k == "Vcpu")
```
done

alindima commented on src/firecracker/src/main.rs:
```diff
@@ -202,16 +206,23 @@ fn main() {
+    if seccomp_enabled {
+        if let Some(path) = arguments.value_as_string("seccomp-filter") {
+            let filter_file =
+                File::open(&path).unwrap_or_else(|err| panic!("Seccomp file error: {}", err));
```
fixed

alindima commented on src/firecracker/src/main.rs:
```diff
@@ -202,16 +206,23 @@ fn main() {
+                panic!("Could not create seccomp filter: {}", err);
+            });
+        }
+    }
```
done

alindima commented on src/seccomp/src/compiler.rs:
```diff
@@ -0,0 +1,561 @@
+        match *self {
+            MultipleSyscallFields => write!(
+                f,
+                "Cannot use both `syscalls` and `syscall` properties in an object"
```
done

alindima commented on src/vmm/src/default_syscalls/filters.rs:
```diff
@@ -169,7 +169,7 @@ pub fn get_default_filters() -> Result<BpfThreadMap, Error> {
         SeccompAction::Trap,
     )?
-    .try_into()?;
+    .into_bpf(std::env::consts::ARCH)?;
```
done

alindima commented on src/firecracker/src/main.rs:
```diff
@@ -108,6 +107,11 @@ fn main() {
+        .arg(
+            Argument::new("no-seccomp")
+                .takes_value(false)
+                .help("Optional parameter which allows starting and using a microVM without seccomp filtering. Not recommended")
```
done

alindima commented on CHANGELOG.md:
```diff
@@ -34,6 +34,8 @@
 - Added devtool test `-m|--cpuset-mems` flag for memory confinement when tests
   run.
+- Added `--no-seccomp` parameter for disabling the default seccomp filters
+- Added `--seccomp-filter` parameter for supplying user-provided, custom filters
```
done

alindima commented on src/firecracker/src/main.rs:
```diff
@@ -246,8 +257,12 @@ fn main() {
     } else {
+        let mut seccomp_filters: BpfThreadMap = seccomp_filters
+            .into_iter()
+            .filter(|(k, _)| k == "Vmm" || k == "Vcpu")
```
done

alindima commented on src/seccomp/src/lib.rs:
```diff
@@ -417,6 +417,12 @@ pub struct SeccompRule {
+/// Type that associates the syscall number to its SeccompRules
+pub type SeccompRuleMap = BTreeMap<i64, Vec<SeccompRule>>;
+
+/// Type that associates a thread category to a BPF program
```
done

alindima commented on src/firecracker/src/main.rs:
```diff
@@ -4,6 +4,7 @@ mod api_server_adapter;
 mod metrics;
 
 use std::fs;
+use std::fs::File;
```
done

alindima commented on src/vmm/src/builder.rs:
```diff
@@ -344,13 +344,14 @@ pub fn build_microvm_for_boot(
 
     // Move vcpus to their own threads and start their state machine in the 'Paused' state.
-    vmm.start_vcpus(vcpus, seccomp_filter).map_err(Internal)?;
+    vmm.start_vcpus(vcpus, seccomp_filters.remove("Vcpu").unwrap())
```
fixed

alindima commented on src/vmm/src/builder.rs:
```diff
@@ -344,13 +344,14 @@ pub fn build_microvm_for_boot(
     // altogether is the desired behaviour.
     // Keep this as the last step before resuming vcpus.
-    SeccompFilter::apply(seccomp_filter.to_vec())
+    SeccompFilter::apply(seccomp_filters.remove("Vmm").unwrap())
```
fixed

alindima commented on src/seccomp/src/lib.rs:
```diff
@@ -1090,13 +1082,15 @@ fn BPF_STMT(code: u16, k: u32) -> sock_filter {
 #[allow(non_snake_case)]
 #[inline(always)]
-fn VALIDATE_ARCHITECTURE() -> Vec<sock_filter> {
+fn VALIDATE_ARCHITECTURE(target_arch: &str) -> Vec<sock_filter> {
```
fixed

alindima commented on src/seccomp/src/syscall_table.rs:
```diff
@@ -0,0 +1,673 @@
+    pub fn new(arch: String) -> Self {
+        let mut instance = Self {
+            arch,
+            map: HashMap::with_capacity(332), // nr of syscalls for x86_64 (upper bound)
```
done

alindima commented on src/seccomp/src/compiler.rs:
```diff
@@ -0,0 +1,561 @@
+//! It also defines most of the objects that a seccomp filter is deserialized into:
+//! `Filter`, `SyscallObject`.
+//!
+//! The rest of objects are deserialized directly into the IR (intermediate representation), declared in `lib.rs`:
```
done

alindima commented on src/seccomp/src/compiler.rs:
```diff
@@ -0,0 +1,561 @@
+//!    (via SeccompFilter.into_bpf(...))
+//!                   |
+//!                   V
+//!     collection of `BpfProgram` objects
```
thanks. updated

alindima commented on src/seccomp/src/bin.rs:
```diff
@@ -1,8 +1,568 @@
+        process::exit(1);
+    });
+
+    compile(&args).unwrap_or_else(|err| panic!("Seccompiler error: {}", err));
```
done

alindima commented on src/vmm/src/default_syscalls/filters.rs:
```diff
@@ -186,21 +186,45 @@ pub fn get_empty_filters() -> BpfThreadMap {
+    /// Filter deserialitaion error
+    Deserialization(DeserializationError),
+    /// Invalid thread categories
+    ThreadCategories,
```
done

alindima commented on src/vmm/src/default_syscalls/filters.rs:
```diff
@@ -186,21 +186,45 @@ pub fn get_empty_filters() -> BpfThreadMap {
+pub fn get_custom_filters(mut file: File) -> Result<BpfThreadMap, FilterError> {
+    let filters = deserialize_binary(&mut file).map_err(FilterError::Deserialization)?;
+
+    if filters.contains_key("Vmm") && filters.contains_key("Api") && filters.contains_key("Vcpu") {
```
done

alindima commented:
I have also just rewritten the seccomp python integration tests in Firecracker to use seccompiler instead of the old seccomp primitives.

iulianbarbu commented on src/seccomp/src/syscall_table.rs:
```diff
@@ -0,0 +1,679 @@
+    use seccomp::TargetArch;
+
+    #[test]
+    fn test_get_syscall_nr() {
```
I think we should test the syscall table here. The expectation is to keep all the known syscalls at any time. If not, this can be detected here.

iulianbarbu commented on src/vmm/src/default_syscalls/filters.rs:
```diff
@@ -1,20 +1,83 @@
+}
+
+/// Return an error if the BpfThreadMap contains invalid thread categories.
+fn filter_thread_categories(map: BpfThreadMap) -> Result<BpfThreadMap, FilterError> {
```
nit:
This seems like more handling than is needed. If the `BpfThreadMap` has entries responsible for other thread categories, or lacks mandatory thread categories, we can simply return that the expected thread categories are `vmm`,`api` and `vcpu`.

I feel that the level of handling here implies logic which sets the expectation for more and more error reporting, which is not needed IMO.

iulianbarbu replied:
This can be as well a BpfThreadMap `validate` method.

iulianbarbu commented on src/vmm/src/default_syscalls/filters.rs:
```diff
@@ -1,20 +1,83 @@
 /// or why the sycall/condition is needed.
-pub fn default_filter() -> Result<SeccompFilter, Error> {
-    Ok(SeccompFilter::new(
+pub fn get_default_filters() -> Result<BpfThreadMap, Error> {
```
An implementation for Default for `BpfThreadMap` would fit better.

iulianbarbu commented on src/vmm/src/default_syscalls/filters.rs:
```diff
@@ -169,32 +232,73 @@ pub fn default_filter() -> Result<SeccompFilter, Error> {
-            .map_err(SeccompError::SeccompFilter),
-    }
+/// Retrieve empty seccomp filters.
+pub fn get_empty_filters() -> BpfThreadMap {
```
Seems more like a `BpfThreadMap` method to me.

iulianbarbu commented on src/vmm/src/default_syscalls/filters.rs:
```diff
@@ -169,32 +232,73 @@ pub fn default_filter() -> Result<SeccompFilter, Error> {
+}
+
+/// Retrieve custom seccomp filters.
+pub fn get_custom_filters(mut file: File) -> Result<BpfThreadMap, FilterError> {
```
ditto

iulianbarbu commented on src/vmm/src/signal_handler.rs:
```diff
@@ -193,31 +192,141 @@ mod tests {
-
-            assert!(SeccompFilter::apply(filter.try_into().unwrap()).is_ok());
+            // Test with a seccomp filter that allows all syscalls, except for `SYS_mkdirat`.
+            // For some reason, directly calling `kick` with SIGSYS, like we do with the
```
Not sure I get this comment. Do you mean directly calling `SYS_kill` with SIGSYS results in an error? SIGSYS is generated by `seccomp` or by issuing a syscall with bad arguments.

iulianbarbu commented on tests/framework/utils.py:
```diff
@@ -359,3 +359,19 @@ def eager_map(func, iterable):
         continue
+
+
+def assert_seccomp_level(pid, seccomp_level):
```
Are we programmatically setting this level somewhere? What are the options?

iulianbarbu reviewed (commented):
This is a really cool feature! Nice job! The comments are just styling related.

aghecenco commented:
LGTM, I think it will be good to merge once the syscall table story is sorted out. Awesome work! :100:

andreeaflorescu commented:
Before merging this we should add a fix for a potential DoS of the host when the BPF file is invalid. As we're using bincode for serializing & deserializing, we should first set a limit to the allocated memory using the bincode configuration options: https://docs.rs/bincode/1.3.1/bincode/config/trait.Options.html#byte-limit-details. The fix should also be accompanied by a test that checks for the allocated memory when loading BPF.

alindima commented:
> Before merging this we should add a fix for a potential DoS of the host when the BPF file is invalid. As we're using bincode for serializing & deserializing, we should first set a limit to the allocated memory using the bincode configuration options: https://docs.rs/bincode/1.3.1/bincode/config/trait.Options.html#byte-limit-details. The fix should also be accompanied by a test that checks for the allocated memory when loading BPF.

Thanks for pointing this out. This PR is on a feature branch (so this bug shouldn't be a problem) and I am planning on opening another PR for this.
The current PR is already getting very big and I don't want to add more things on top of it

alindima commented on src/seccomp/src/lib.rs:
```diff
@@ -1665,16 +1739,7 @@ mod tests {
-                (4, 0)
-            }
-        };
+        let (msb_offset, lsb_offset) = { (4, 0) };
```
removed the comment

alindima commented on src/seccomp/src/lib.rs:
```diff
@@ -1709,16 +1774,7 @@ mod tests {
-                (4, 0)
-            }
-        };
+        let (msb_offset, lsb_offset) = { (4, 0) };
```
removed the comment

alindima commented:
Added devtool command for generating syscall tables

alindima commented:
Removed the option of having an object with a `syscalls` property. This introduced some unneccessary code complexity, with little benefit.
Addressed comments and ready for another review, PTAL

alindima commented on src/seccomp/src/compiler.rs:
```diff
@@ -0,0 +1,600 @@
+    }
+
+    /// Transforms the deserialized `Filter` into a `SeccompFilter` (IR language).
+    fn make_seccomp_filter(&self, filter: Filter) -> Result<SeccompFilter> {
```
We synced offline and decided that this is not a good idea

alindima commented on src/seccomp/src/compiler.rs:
```diff
@@ -0,0 +1,600 @@
+            .map(|(_, filter)| filter.validate())
+            .find(|result| result.is_err())
+            .or_else(|| Some(Ok(())))
+            .unwrap()
```
Rewritten this to use `filter_map`, and got rid of the `unwrap`

iulianbarbu commented on src/seccomp/src/compiler.rs:
```diff
@@ -0,0 +1,420 @@
+                .get_syscall_nr(&syscall_name)
+                .ok_or_else(|| Error::SyscallName(syscall_name.clone(), self.arch))?;
+            let rule_accumulator = rule_map.entry(syscall_nr).or_insert_with(|| vec![]);
+            let conditions = syscall_rule.conditions.or_else(|| Some(vec![])).unwrap();
+
+            rule_accumulator.push(SeccompRule::new(conditions, action));
```
nit:
```suggestion
            match syscall_rule.conditions {
                Some(conditions) =>  rule_accumulator.push(SeccompRule::new(conditions, action));
                None => rule_accumulator.push(SeccompRule::new(vec![], action));
            }
```

iulianbarbu commented on src/seccomp/src/lib.rs:
```diff
@@ -880,6 +984,16 @@ impl SeccompFilter {
+    }
+
+    /// Creates an empty `SeccompFilter` which allows everything.
+    pub fn empty(target_arch: &str) -> Result<Self> {
```
nit: Can this be a `Default` implementation? Do you think it makes sense from a semantic pov?

iulianbarbu commented on src/seccomp/src/syscall_table/mod.rs:
```diff
@@ -0,0 +1,64 @@
+}
+
+/// Number of syscalls for x86_64 (upper bound).
+const MAP_CAPACITY: usize = 351;
```
Should this be autogenerated as well?

iulianbarbu commented on tools/devtool:
```diff
@@ -302,6 +308,20 @@ validate_version() {
+# Validate the user supplied kernel version number.
+# It must be composed of 2 groups of integers separated by dot, with an optional third group.
+validate_kernel_version() {
+    declare version_regex="^([0-9]+.)[0-9]+(.[0-9]+)?$"
```
nit: We can use `local` to denote that the variable is used only within this scope. `declare` is not really necessary.

iulianbarbu reviewed (approved):
Really awesome job, @alindima !

aghecenco commented on src/seccomp/src/lib.rs:
```diff
@@ -1116,64 +1212,58 @@ fn EXAMINE_SYSCALL() -> Vec<sock_filter> {
+    reader: &mut dyn Read,
+) -> std::result::Result<BpfThreadMap, DeserializationError> {
+    Ok(
+        bincode::deserialize_from::<&mut dyn Read, BpfThreadMap>(reader)
```
I'm still concerned that this allows unlimited reading into memory, if presented with a huge file to read from.

aghecenco reviewed (approved):
Good to go! Awesome work! :rocket:
